In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('test2.csv')
df.head()

IsCanceled  LeadTime  ArrivalDateYear ArrivalDateMonth  \
0           0       342             2015             July   
1           0       737             2015             July   
2           0         7             2015             July   
3           0        13             2015             July   
4           0        14             2015             July   

   ArrivalDateWeekNumber  ArrivalDateDayOfMonth  StaysInWeekendNights  \
0                     27                      1                     0   
1                     27                      1                     0   
2                     27                      1                     0   
3                     27                      1                     0   
4                     27                      1                     0   

   StaysInWeekNights  Adults  Children  ...      DepositType        Agent  \
0                  0       2         0  ...  No Deposit              NULL   
1                  0       2         0  ...  No Deposit              NULL   
2                  1       1         0  ...  No Deposit              NULL   
3                  1       1         0  ...  No Deposit               304   
4                  2       2         0  ...  No Deposit               240   

       Company DaysInWaitingList CustomerType   ADR  RequiredCarParkingSpaces  \
0         NULL                 0    Transient   0.0                         0   
1         NULL                 0    Transient   0.0                         0   
2         NULL                 0    Transient  75.0                         0   
3         NULL                 0    Transient  75.0                         0   
4         NULL                 0    Transient  98.0                         0   

   TotalOfSpecialRequests ReservationStatus ReservationStatusDate  
0                       0         Check-Out            2015-07-01  
1                       0         Check-Out            2015-07-01  
2                       0         Check-Out            2015-07-02  
3                       0         Check-Out            2015-07-02  
4                       1         Check-Out            2015-07-03  

[5 rows x 31 columns]

In [3]:
df.T

0                 1      \
IsCanceled                                  0                 0   
LeadTime                                  342               737   
ArrivalDateYear                          2015              2015   
ArrivalDateMonth                         July              July   
ArrivalDateWeekNumber                      27                27   
ArrivalDateDayOfMonth                       1                 1   
StaysInWeekendNights                        0                 0   
StaysInWeekNights                           0                 0   
Adults                                      2                 2   
Children                                    0                 0   
Babies                                      0                 0   
Meal                                BB                BB          
Country                                   PRT               PRT   
MarketSegment                          Direct            Direct   
DistributionChannel                    Direct            Direct   
IsRepeatedGuest                             0                 0   
PreviousCancellations                       0                 0   
PreviousBookingsNotCanceled                 0                 0   
ReservedRoomType             C                 C                  
AssignedRoomType             C                 C                  
BookingChanges                              3                 4   
DepositType                   No Deposit        No Deposit        
Agent                                    NULL              NULL   
Company                                  NULL              NULL   
DaysInWaitingList                           0                 0   
CustomerType                        Transient         Transient   
ADR                                       0.0               0.0   
RequiredCarParkingSpaces                    0                 0   
TotalOfSpecialRequests                      0                 0   
ReservationStatus                   Check-Out         Check-Out   
ReservationStatusDate              2015-07-01        2015-07-01   

                                        2                 3      \
IsCanceled                                  0                 0   
LeadTime                                    7                13   
ArrivalDateYear                          2015              2015   
ArrivalDateMonth                         July              July   
ArrivalDateWeekNumber                      27                27   
ArrivalDateDayOfMonth                       1                 1   
StaysInWeekendNights                        0                 0   
StaysInWeekNights                           1                 1   
Adults                                      1                 1   
Children                                    0                 0   
Babies                                      0                 0   
Meal                                BB                BB          
Country                                   GBR               GBR   
MarketSegment                          Direct         Corporate   
DistributionChannel                    Direct         Corporate   
IsRepeatedGuest                             0                 0   
PreviousCancellations                       0                 0   
PreviousBookingsNotCanceled                 0                 0   
ReservedRoomType             A                 A                  
AssignedRoomType             C                 A                  
BookingChanges                              0                 0   
DepositType                   No Deposit        No Deposit        
Agent                                    NULL               304   
Company                                  NULL              NULL   
DaysInWaitingList                           0                 0   
CustomerType                        Transient         Transient   
ADR                                      75.0              75.0   
RequiredCarParkingS

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40060 entries, 0 to 40059
Data columns (total 31 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   IsCanceled                   40060 non-null  int64  
 1   LeadTime                     40060 non-null  int64  
 2   ArrivalDateYear              40060 non-null  int64  
 3   ArrivalDateMonth             40060 non-null  object 
 4   ArrivalDateWeekNumber        40060 non-null  int64  
 5   ArrivalDateDayOfMonth        40060 non-null  int64  
 6   StaysInWeekendNights         40060 non-null  int64  
 7   StaysInWeekNights            40060 non-null  int64  
 8   Adults                       40060 non-null  int64  
 9   Children                     40060 non-null  int64  
 10  Babies                       40060 non-null  int64  
 11  Meal                         40060 non-null  object 
 12  Country                      39596 non-null  object 
 13  MarketSegment   

In [5]:
df.describe().T

count         mean        std      min     25%  \
IsCanceled                   40060.0     0.277634   0.447837     0.00     0.0   
LeadTime                     40060.0    92.675686  97.285315     0.00    10.0   
ArrivalDateYear              40060.0  2016.121443   0.722347  2015.00  2016.0   
ArrivalDateWeekNumber        40060.0    27.140864  14.005441     1.00    16.0   
ArrivalDateDayOfMonth        40060.0    15.821243   8.883708     1.00     8.0   
StaysInWeekendNights         40060.0     1.189815   1.147812     0.00     0.0   
StaysInWeekNights            40060.0     3.128732   2.461329     0.00     1.0   
Adults                       40060.0     1.867149   0.697285     0.00     2.0   
Children                     40060.0     0.128682   0.445195     0.00     0.0   
Babies                       40060.0     0.013904   0.118998     0.00     0.0   
IsRepeatedGuest              40060.0     0.044383   0.205948     0.00     0.0   
PreviousCancellations        40060.0     0.101722   1.335115     0.00     0.0   
PreviousBookingsNotCanceled  40060.0     0.146455   1.001955     0.00     0.0   
BookingChanges               40060.0     0.287968   0.726548     0.00     0.0   
DaysInWaitingList            40060.0     0.527758   7.428580     0.00     0.0   
ADR                          40060.0    94.952930  61.442418    -6.38    50.0   
RequiredCarParkingSpaces     40060.0     0.138068   0.351001     0.00     0.0   
TotalOfSpecialRequests       40060.0     0.619770   0.813930     0.00     0.0   

                                50%     75%     max  
IsCanceled                      0.0     1.0     1.0  
LeadTime                       57.0   155.0   737.0  
ArrivalDateYear              2016.0  2017.0  2017.0  
ArrivalDateWeekNumber          28.0    38.0    53.0  
ArrivalDateDayOfMonth          16.0    24.0    31.0  
StaysInWeekendNights            1.0     2.0    19.0  
StaysInWeekNights               3.0     5.0    50.0  
Adults                          2.0     2.0    55.0  
Children                        0.0     0.0    10.0  
Babies                          0.0     0.0     2.0  
IsRepeatedGuest                 0.0     0.0     1.0  
PreviousCancellations           0.0     0.0    26.0  
PreviousBookingsNotCanceled     0.0     0.0    30.0  
BookingChanges                  0.0     0.0    17.0  
DaysInWaitingList               0.0     0.0   185.0  
ADR                            75.0   125.0   508.0  
RequiredCarParkingSpaces        0.0     0.0     8.0  
TotalOfSpecialRequests          0.0     1.0     5.0

# 데이터셋
- 출처 : 포르투칼 남부 algrabe지역 호텔의 자산관리 시스템(PMS: Prperty Management System)
- 분야 : 숙박업(호텔)

## Business Understanding
- 어떤 분야인가? 숙박업(호텔) -> 최대한 많은 예약을 체크인 하여 비어있는 방을 최소화하는 것을 목표로 함.
- 많은 예약을 받는다.
- 예약 취소를 최소화 한다. --> 왜 취소하는가? 기존 고객인가? 어떤 경로로 예약을 했는가? 어떤 경로의 고객이 취소를 가장 많이 하는가?
- 밥, 부가 서비스, 여행

In [7]:
df.columns

Index(['IsCanceled', 'LeadTime', 'ArrivalDateYear', 'ArrivalDateMonth',
       'ArrivalDateWeekNumber', 'ArrivalDateDayOfMonth',
       'StaysInWeekendNights', 'StaysInWeekNights', 'Adults', 'Children',
       'Babies', 'Meal', 'Country', 'MarketSegment', 'DistributionChannel',
       'IsRepeatedGuest', 'PreviousCancellations',
       'PreviousBookingsNotCanceled', 'ReservedRoomType', 'AssignedRoomType',
       'BookingChanges', 'DepositType', 'Agent', 'Company',
       'DaysInWaitingList', 'CustomerType', 'ADR', 'RequiredCarParkingSpaces',
       'TotalOfSpecialRequests', 'ReservationStatus', 'ReservationStatusDate'],
      dtype='object')

In [8]:
set(df['DistributionChannel']) 
# BB – Bed & Breakfast;
# HB – Half board (breakfast and one other meal – usually dinner);
# FB – Full board (breakfast, lunch and dinner)

{'Corporate', 'Direct', 'TA/TO', 'Undefined'}

1. 평균적인 사전예약 기간
2. 예약된 방과 전환되는 방 사이의 선호도 -> 직접예약과의 관계?
3. 직접 예약과 대리를 통한 예약 사이의 차이 -> 취소, 방전환, 고객의 나이
4. 재방문 하는 사람들의 방의 변화, 식사량
5. 신규고객 분석.
6. 방에 대한 정보????

- check in ~ check out 까지
- 가능한한 모든 객실이 full-booking되어야 함
- 수요량 예측 -> 예약 취소확률 -> 오버 부킹 관리
- 예약 취소에 영향을 주는 요소를 찾아냄 -> 시각화를 통해 경영에 반영할 수 있게함
- 유사한 분야 : 비행기 예약, 식당 예약, 영화 관람, 

## Data Understanding

In [12]:
df.columns

Index(['IsCanceled', 'LeadTime', 'ArrivalDateYear', 'ArrivalDateMonth',
       'ArrivalDateWeekNumber', 'ArrivalDateDayOfMonth',
       'StaysInWeekendNights', 'StaysInWeekNights', 'Adults', 'Children',
       'Babies', 'Meal', 'Country', 'MarketSegment', 'DistributionChannel',
       'IsRepeatedGuest', 'PreviousCancellations',
       'PreviousBookingsNotCanceled', 'ReservedRoomType', 'AssignedRoomType',
       'BookingChanges', 'DepositType', 'Agent', 'Company',
       'DaysInWaitingList', 'CustomerType', 'ADR', 'RequiredCarParkingSpaces',
       'TotalOfSpecialRequests', 'ReservationStatus', 'ReservationStatusDate'],
      dtype='object')

In [13]:
df.T

0                 1      \
IsCanceled                                  0                 0   
LeadTime                                  342               737   
ArrivalDateYear                          2015              2015   
ArrivalDateMonth                         July              July   
ArrivalDateWeekNumber                      27                27   
ArrivalDateDayOfMonth                       1                 1   
StaysInWeekendNights                        0                 0   
StaysInWeekNights                           0                 0   
Adults                                      2                 2   
Children                                    0                 0   
Babies                                      0                 0   
Meal                                BB                BB          
Country                                   PRT               PRT   
MarketSegment                          Direct            Direct   
DistributionChannel                    Direct            Direct   
IsRepeatedGuest                             0                 0   
PreviousCancellations                       0                 0   
PreviousBookingsNotCanceled                 0                 0   
ReservedRoomType             C                 C                  
AssignedRoomType             C                 C                  
BookingChanges                              3                 4   
DepositType                   No Deposit        No Deposit        
Agent                                    NULL              NULL   
Company                                  NULL              NULL   
DaysInWaitingList                           0                 0   
CustomerType                        Transient         Transient   
ADR                                       0.0               0.0   
RequiredCarParkingSpaces                    0                 0   
TotalOfSpecialRequests                      0                 0   
ReservationStatus                   Check-Out         Check-Out   
ReservationStatusDate              2015-07-01        2015-07-01   

                                        2                 3      \
IsCanceled                                  0                 0   
LeadTime                                    7                13   
ArrivalDateYear                          2015              2015   
ArrivalDateMonth                         July              July   
ArrivalDateWeekNumber                      27                27   
ArrivalDateDayOfMonth                       1                 1   
StaysInWeekendNights                        0                 0   
StaysInWeekNights                           1                 1   
Adults                                      1                 1   
Children                                    0                 0   
Babies                                      0                 0   
Meal                                BB                BB          
Country                                   GBR               GBR   
MarketSegment                          Direct         Corporate   
DistributionChannel                    Direct         Corporate   
IsRepeatedGuest                             0                 0   
PreviousCancellations                       0                 0   
PreviousBookingsNotCanceled                 0                 0   
ReservedRoomType             A                 A                  
AssignedRoomType             C                 A                  
BookingChanges                              0                 0   
DepositType                   No Deposit        No Deposit        
Agent                                    NULL               304   
Company                                  NULL              NULL   
DaysInWaitingList                           0                 0   
CustomerType                        Transient         Transient   
ADR                                      75.0              75.0   
RequiredCarParkingS

### 경우의 수 컬럼을 처리해야함
- 경우의 수 -> 숫자로 변환
- 컴퓨터는 학습할 때 데이터가 숫자여야 처리가 쉬움, 대부분 그러함
- 방법 : Apply Domain Knowedge, Frequency Encoding, Target Encoding, One-hot Encoding

1. Apply Domain Knowledge
2. One-hot Encoding
3. Frequency Encoding  -권장하지 않음
4. Target Encodeing    -권장하지 않음

#### Apply Domain Knowledge
- 이 방식을 가장 먼저 고려
- 가장 이상적이고 우선적 적용
- 기준을 정했을 때, 정도의 차이가 발생
- 순서를 매기면 의미가 있는 경우

In [16]:
set(df['Country'])

{'AGO',
 'ALB',
 'AND',
 'ARE',
 'ARG',
 'ARM',
 'AUS',
 'AUT',
 'AZE',
 'BDI',
 'BEL',
 'BGR',
 'BHR',
 'BHS',
 'BIH',
 'BLR',
 'BRA',
 'BWA',
 'CAF',
 'CHE',
 'CHL',
 'CHN',
 'CIV',
 'CMR',
 'CN',
 'COL',
 'COM',
 'CPV',
 'CRI',
 'CUB',
 'CYM',
 'CYP',
 'CZE',
 'DEU',
 'DJI',
 'DNK',
 'DOM',
 'DZA',
 'ECU',
 'EGY',
 'ESP',
 'EST',
 'FIN',
 'FJI',
 'FRA',
 'GBR',
 'GEO',
 'GGY',
 'GIB',
 'GRC',
 'HKG',
 'HRV',
 'HUN',
 'IDN',
 'IND',
 'IRL',
 'IRN',
 'ISL',
 'ISR',
 'ITA',
 'JAM',
 'JEY',
 'JOR',
 'JPN',
 'KAZ',
 'KOR',
 'KWT',
 'LBN',
 'LKA',
 'LTU',
 'LUX',
 'LVA',
 'MAC',
 'MAR',
 'MDG',
 'MDV',
 'MEX',
 'MKD',
 'MLT',
 'MOZ',
 'MUS',
 'MWI',
 'MYS',
 'NGA',
 'NLD',
 'NOR',
 'NPL',
 'NZL',
 'OMN',
 'PAK',
 'PER',
 'PHL',
 'PLW',
 'POL',
 'PRI',
 'PRT',
 'QAT',
 'ROU',
 'RUS',
 'SAU',
 'SEN',
 'SGP',
 'SMR',
 'SRB',
 'SUR',
 'SVK',
 'SVN',
 'SWE',
 'SYC',
 'SYR',
 'TGO',
 'THA',
 'TUN',
 'TUR',
 'TWN',
 'UGA',
 'UKR',
 'URY',
 'USA',
 'UZB',
 'VEN',
 'VNM',
 'ZAF',
 'ZMB',
 'ZWE',
 

In [17]:
mapper = {
    'SC       ' : 0,
    'Undefined' : 0,
    'BB       ' : 1,
    'HB       ' : 2,
    'FB       ' : 3
}

df['day_stay'] = df['Meal'].map(mapper)

In [18]:
df[['Meal', 'day_stay']].head(20).T 

0          1          2          3          4          5   \
Meal      BB         BB         BB         BB         BB         BB          
day_stay          1          1          1          1          1          1   

                 6          7          8          9          10         11  \
Meal      BB         FB         BB         HB         BB         HB          
day_stay          1          3          1          2          1          2   

                 12         13         14         15         16         17  \
Meal      BB         HB         BB         BB         BB         BB          
day_stay          1          2          1          1          1          1   

                 18         19  
Meal      BB         BB         
day_stay          1          1

In [19]:
df['foreigner'] = 1
df[df['Country'] == 'PRT']['foreigner'] = 0

C:\Users\405-11\AppData\Local\Temp\ipykernel_276\3004285217.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[df['Country'] == 'PRT']['foreigner'] = 0


In [20]:
df.loc[df['Country'] == 'PRT','foreigner'] = 0

In [21]:
df[df['Country'] == 'PRT']

IsCanceled  LeadTime  ArrivalDateYear ArrivalDateMonth  \
0               0       342             2015             July   
1               0       737             2015             July   
6               0         0             2015             July   
7               0         9             2015             July   
8               1        85             2015             July   
...           ...       ...              ...              ...   
40024           0        41             2017           August   
40025           0       161             2017           August   
40026           0       206             2017           August   
40028           0        43             2017           August   
40050           0       173             2017           August   

       ArrivalDateWeekNumber  ArrivalDateDayOfMonth  StaysInWeekendNights  \
0                         27                      1                     0   
1                         27                      1                     0   
6                         27                      1                     0   
7                         27                      1                     0   
8                         27                      1                     0   
...                      ...                    ...                   ...   
40024                     35                     29                     2   
40025                     35                     30                     2   
40026                     35                     31                     2   
40028                     35                     31                     2   
40050                     35                     27                     4   

       StaysInWeekNights  Adults  Children  ...      Company  \
0                      0       2         0  ...         NULL   
1                      0       2         0  ...         NULL   
6                      2       2         0  ...         NULL   
7                      2       2         0  ...         NULL   
8                      3       2         0  ...         NULL   
...                  ...     ...       ...  ...          ...   
40024                  6       2         0  ...         NULL   
40025                  5       2         0  ...         NULL   
40026                  4       3         0  ...         NULL   
40028                  4       1         0  ...         NULL   
40050                  9       2         0  ...         NULL   

      DaysInWaitingList     CustomerType     ADR RequiredCarParkingSpaces  \
0                     0        Transient    0.00                        0   
1                     0        Transient    0.00                        0   
6                     0        Transient  107.00                        0   
7                     0        Transient  103.00                        0   
8                     0        Transient   82.00                        0   
...                 ...              ...     ...                      ...   
40024                 0        Transient  177.50                        0   
40025                 0        Transient  213.43                        0   
40026                 0        Transient  209.67                        1   
40028                 0        Transient  158.53                        0   
40050                 0  Transient-Party  168.92                        0   

       TotalOfSpecialRequests  ReservationStatus  ReservationStatusDate  \
0                           0          Check-Out             2015-07-01   
1                           0          Check-Out             2015-07-01   
6                           0          Check-Out             2015-07-03   
7                           1          Check-Out             2015-07-03   
8                           1           Canceled             2015-05-06   
...                       ...                ...                    ...   
40024                       2          Check-Out             2017-09-06   
40025  

In [22]:
df.T


0                 1      \
IsCanceled                                  0                 0   
LeadTime                                  342               737   
ArrivalDateYear                          2015              2015   
ArrivalDateMonth                         July              July   
ArrivalDateWeekNumber                      27                27   
ArrivalDateDayOfMonth                       1                 1   
StaysInWeekendNights                        0                 0   
StaysInWeekNights                           0                 0   
Adults                                      2                 2   
Children                                    0                 0   
Babies                                      0                 0   
Meal                                BB                BB          
Country                                   PRT               PRT   
MarketSegment                          Direct            Direct   
DistributionChannel                    Direct            Direct   
IsRepeatedGuest                             0                 0   
PreviousCancellations                       0                 0   
PreviousBookingsNotCanceled                 0                 0   
ReservedRoomType             C                 C                  
AssignedRoomType             C                 C                  
BookingChanges                              3                 4   
DepositType                   No Deposit        No Deposit        
Agent                                    NULL              NULL   
Company                                  NULL              NULL   
DaysInWaitingList                           0                 0   
CustomerType                        Transient         Transient   
ADR                                       0.0               0.0   
RequiredCarParkingSpaces                    0                 0   
TotalOfSpecialRequests                      0                 0   
ReservationStatus                   Check-Out         Check-Out   
ReservationStatusDate              2015-07-01        2015-07-01   
day_stay                                    1                 1   
foreigner                                   0                 0   

                                        2                 3      \
IsCanceled                                  0                 0   
LeadTime                                    7                13   
ArrivalDateYear                          2015              2015   
ArrivalDateMonth                         July              July   
ArrivalDateWeekNumber                      27                27   
ArrivalDateDayOfMonth                       1                 1   
StaysInWeekendNights                        0                 0   
StaysInWeekNights                           1                 1   
Adults                                      1                 1   
Children                                    0                 0   
Babies                                      0                 0   
Meal                                BB                BB          
Country                                   GBR               GBR   
MarketSegment                          Direct         Corporate   
DistributionChannel                    Direct         Corporate   
IsRepeatedGuest                             0                 0   
PreviousCancellations                       0                 0   
PreviousBookingsNotCanceled                 0                 0   
ReservedRoomType             A                 A                  
AssignedRoomType             C                 A                  
BookingChanges                              0                 0   
DepositType                   No Deposit        No Deposit        
Agent                                    NULL               304   
Company                                  NULL              NULL   
DaysInWaitingList                           0                 0   
CustomerType       

#### Frequency Encoding
- 경우의 수를 빈도(%)로 변환
- 각 경우의 수가 데이터셋에서 차지하는 비율을 적용해보자

In [24]:
df['Country'].value_counts()

Country
PRT    17630
GBR     6814
ESP     3957
IRL     2166
FRA     1611
       ...  
BIH        1
MUS        1
COM        1
UGA        1
DJI        1
Name: count, Length: 125, dtype: int64

In [25]:
cr = pd.DataFrame(df['Country'].value_counts(normalize=True)).reset_index()

In [26]:
cr

Country  proportion
0       PRT    0.445247
1       GBR    0.172088
2       ESP    0.099934
3       IRL    0.054702
4       FRA    0.040686
..      ...         ...
120     BIH    0.000025
121     MUS    0.000025
122     COM    0.000025
123     UGA    0.000025
124     DJI    0.000025

[125 rows x 2 columns]

In [27]:
df = pd.merge(df, cr)

In [28]:
df

IsCanceled  LeadTime  ArrivalDateYear ArrivalDateMonth  \
0               0       342             2015             July   
1               0       737             2015             July   
2               0         0             2015             July   
3               0         9             2015             July   
4               1        85             2015             July   
...           ...       ...              ...              ...   
39591           0       130             2017             July   
39592           0       130             2017             July   
39593           0       130             2017             July   
39594           0       130             2017             July   
39595           0         0             2017             July   

       ArrivalDateWeekNumber  ArrivalDateDayOfMonth  StaysInWeekendNights  \
0                         27                      1                     0   
1                         27                      1                     0   
2                         27                      1                     0   
3                         27                      1                     0   
4                         27                      1                     0   
...                      ...                    ...                   ...   
39591                     28                      9                     2   
39592                     28                      9                     2   
39593                     28                      9                     2   
39594                     28                      9                     2   
39595                     29                     18                     0   

       StaysInWeekNights  Adults  Children  ...  DaysInWaitingList  \
0                      0       2         0  ...                  0   
1                      0       2         0  ...                  0   
2                      2       2         0  ...                  0   
3                      2       2         0  ...                  0   
4                      3       2         0  ...                  0   
...                  ...     ...       ...  ...                ...   
39591                  0       2         0  ...                  0   
39592                  0       2         0  ...                  0   
39593                  0       2         0  ...                  0   
39594                  0       1         0  ...                  0   
39595                  1       2         2  ...                  0   

          CustomerType    ADR RequiredCarParkingSpaces TotalOfSpecialRequests  \
0            Transient    0.0                        0                      0   
1            Transient    0.0                        0                      0   
2            Transient  107.0                        0                      0   
3            Transient  103.0                        0                      1   
4            Transient   82.0                        0                      1   
...                ...    ...                      ...                    ...   
39591  Transient-Party  160.0                        0                      1   
39592  Transient-Party  160.0                        0                      1   
39593  Transient-Party  178.0                        0                      1   
39594  Transient-Party  160.0                        0                      1   
39595        Transient  273.0                        0                      0   

       ReservationStatus  ReservationStatusDate  day_stay foreigner proportion  
0              Check-Out             2015-07-01         1         0   0.445247  
1              Check-Out             2015-07-01         1         0   0.445247  
2              Check-Out             2015-07-03         1         0   0.445247  
3              Check-Out             2015-07-03         3         0   0.445247  
4               Canceled             2015-05-06         1         0   0.445247  
.

#### Target Encoding
- Label(Target) 정보를 경우의 수 별로 적용
- Label이 있어야만 가능
- 데이터에 선입견이 들어갈 수 있다 : 과적합의 가능성이 늘어남


In [30]:
import numpy as np

In [31]:
df.groupby('Country').agg({'IsCanceled' : np.mean}).reset_index().sort_values(by='IsCanceled')
# 취소확률이 0인 Country인 사람들은 취소를 하지 않는다? 라는 해석이 나올수 있다.(그냥 빈도가 낮아서 그런것임)
# 빈도수가 높은 데이터의 경우에는 필요에 따라 고정관념으로 해줄 필요성이 있을수도 있음

C:\Users\405-11\AppData\Local\Temp\ipykernel_276\1577479290.py:1: FutureWarning: The provided callable <function mean at 0x000001EB6D2C4FE0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df.groupby('Country').agg({'IsCanceled' : np.mean}).reset_index().sort_values(by='IsCanceled')


Country  IsCanceled
62      JOR         0.0
86      NPL         0.0
82      MYS         0.0
81      MWI         0.0
80      MUS         0.0
..      ...         ...
12      BHR         1.0
61      JEY         1.0
100     SEN         1.0
43      FJI         1.0
108     SYC         1.0

[125 rows x 2 columns]

#### One-hot Encoding
- 경우의 수 별로 column을 새로 만들어서 해당 경우의 수이면 1, 아니면 0
- 한 row에서 하나의 column만 hot(1)하게 함
- 일반적으로 도메인 지식을 적용할 수 없는 경우에 사용
- 경우의 수 column 처리의 FM

In [33]:
pd.get_dummies(df['Meal'], dtype=int)

BB         FB         HB         SC         Undefined
0              1          0          0          0          0
1              1          0          0          0          0
2              1          0          0          0          0
3              0          1          0          0          0
4              1          0          0          0          0
...          ...        ...        ...        ...        ...
39591          1          0          0          0          0
39592          1          0          0          0          0
39593          1          0          0          0          0
39594          1          0          0          0          0
39595          1          0          0          0          0

[39596 rows x 5 columns]

In [34]:
pd.get_dummies(df['Country'], dtype=int)

AGO  ALB  AND  ARE  ARG  ARM  AUS  AUT  AZE  BDI  ...  UGA  UKR  URY  \
0        0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
1        0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
2        0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
3        0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
4        0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
39591    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
39592    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
39593    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
39594    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   
39595    0    0    0    0    0    0    0    0    0    0  ...    0    0    0   

       USA  UZB  VEN  VNM  ZAF  ZMB  ZWE  
0        0    0    0    0    0    0    0  
1        0    0    0    0    0    0    0  
2        0    0    0    0    0    0    0  
3        0    0    0    0    0    0    0  
4        0    0    0    0    0    0    0  
...    ...  ...  ...  ...  ...  ...  ...  
39591    0    0    0    0    0    0    0  
39592    0    0    0    0    0    0    0  
39593    0    0    0    0    0    0    0  
39594    0    0    0    0    0    0    0  
39595    0    0    0    0    0    0    0  

[39596 rows x 125 columns]

In [61]:
sum(cr[:5]['proportion'])

0.8126578442266895

- 125개국 데이터를 다 쓰지말고 상위 5개국과 나머지로 나눠 6개의 값으로 다시 설정하자.

In [64]:
sum(cr[-100:]['proportion'])

0.020507121931508172

In [72]:
top_5 = cr[:5]['Country'].values
top_5

array(['PRT', 'GBR', 'ESP', 'IRL', 'FRA'], dtype=object)

In [74]:
def country_hot(country):
    if country in top_5 : 
        return country
    else :
        return 'other'
    

In [80]:
sample = ['PRT', 'GBR', 'ESP', 'KOR','JPN']

for country in sample:
    print(country_hot(country))

PRT
GBR
ESP
other
other


In [82]:
df['Country_simple'] = df['Country'].apply(lambda x: country_hot(x))
pd.get_dummies(df['Country_simple'], dtype=int)

ESP  FRA  GBR  IRL  PRT  other
0        0    0    0    0    1      0
1        0    0    0    0    1      0
2        0    0    0    0    1      0
3        0    0    0    0    1      0
4        0    0    0    0    1      0
...    ...  ...  ...  ...  ...    ...
39591    0    0    0    0    0      1
39592    0    0    0    0    0      1
39593    0    0    0    0    0      1
39594    0    0    0    0    0      1
39595    0    0    0    0    0      1

[39596 rows x 6 columns]

In [86]:
# 참 if 조건 else 거짓  :    1줄 if
df['Country_simple'] = df['Country'].apply(lambda x: x if x in top_5 else 'OTHER')
pd.get_dummies(df['Country_simple'], dtype=int)

ESP  FRA  GBR  IRL  OTHER  PRT
0        0    0    0    0      0    1
1        0    0    0    0      0    1
2        0    0    0    0      0    1
3        0    0    0    0      0    1
4        0    0    0    0      0    1
...    ...  ...  ...  ...    ...  ...
39591    0    0    0    0      1    0
39592    0    0    0    0      1    0
39593    0    0    0    0      1    0
39594    0    0    0    0      1    0
39595    0    0    0    0      1    0

[39596 rows x 6 columns]

- Q) 차를 가져오는 경우와 그렇지 않은 경우 car_yes로 표현
- Q2) 실제로 고객이 예약한 객실과 배정된 객실의 차이가 있는가? "RA_changed"

In [90]:
df['car_yes'] = df['RequiredCarParkingSpaces'].apply(lambda x: True if x > 0 else False)

In [96]:
set(df['car_yes'])

{False, True}

In [104]:

RA_changed = []
for i in range(len(df)) : 
    RA_changed.append(False if df.iloc[i]['ReservedRoomType'] == df.iloc[i]['AssignedRoomType'] else True)

df['RA_changed'] = RA_changed

In [114]:
df[df['IsCanceled']==1].T


4                 5      \
IsCanceled                                  1                 1   
LeadTime                                   85                75   
ArrivalDateYear                          2015              2015   
ArrivalDateMonth                         July              July   
ArrivalDateWeekNumber                      27                27   
ArrivalDateDayOfMonth                       1                 1   
StaysInWeekendNights                        0                 0   
StaysInWeekNights                           3                 3   
Adults                                      2                 2   
Children                                    0                 0   
Babies                                      0                 0   
Meal                                BB                HB          
Country                                   PRT               PRT   
MarketSegment                       Online TA     Offline TA/TO   
DistributionChannel                     TA/TO             TA/TO   
IsRepeatedGuest                             0                 0   
PreviousCancellations                       0                 0   
PreviousBookingsNotCanceled                 0                 0   
ReservedRoomType             A                 D                  
AssignedRoomType             A                 D                  
BookingChanges                              0                 0   
DepositType                   No Deposit        No Deposit        
Agent                                     240                15   
Company                                  NULL              NULL   
DaysInWaitingList                           0                 0   
CustomerType                        Transient         Transient   
ADR                                      82.0             105.5   
RequiredCarParkingSpaces                    0                 0   
TotalOfSpecialRequests                      1                 0   
ReservationStatus                    Canceled          Canceled   
ReservationStatusDate              2015-05-06        2015-04-22   
day_stay                                    1                 2   
foreigner                                   0                 0   
proportion                           0.445247          0.445247   
Country_simple                            PRT               PRT   
car_yes                                 False             False   
RA_changed                              False             False   

                                        6                 14     \
IsCanceled                                  1                 1   
LeadTime                                   23                60   
ArrivalDateYear                          2015              2015   
ArrivalDateMonth                         July              July   
ArrivalDateWeekNumber                      27                27   
ArrivalDateDayOfMonth                       1                 1   
StaysInWeekendNights                        0                 2   
StaysInWeekNights                           4                 5   
Adults                                      2                 2   
Children                                    0                 0   
Babies                                      0                 0   
Meal                                BB                BB          
Country                                   PRT               PRT   
MarketSegment                       Online TA         Online TA   
DistributionChannel                     TA/TO             TA/TO   
IsRepeatedGuest                             0                 0   
PreviousCancellations                       0                 0   
PreviousBookingsNotCanceled                 0                 0   
ReservedRoomType             E                 E                  
AssignedRoomType             E                 E                  
BookingChanges                              0                 0   
DepositType        

In [116]:
royal_customer = df[df['PreviousBookingsNotCanceled'] > 0]

In [122]:
df.columns

Index(['IsCanceled', 'LeadTime', 'ArrivalDateYear', 'ArrivalDateMonth',
       'ArrivalDateWeekNumber', 'ArrivalDateDayOfMonth',
       'StaysInWeekendNights', 'StaysInWeekNights', 'Adults', 'Children',
       'Babies', 'Meal', 'Country', 'MarketSegment', 'DistributionChannel',
       'IsRepeatedGuest', 'PreviousCancellations',
       'PreviousBookingsNotCanceled', 'ReservedRoomType', 'AssignedRoomType',
       'BookingChanges', 'DepositType', 'Agent', 'Company',
       'DaysInWaitingList', 'CustomerType', 'ADR', 'RequiredCarParkingSpaces',
       'TotalOfSpecialRequests', 'ReservationStatus', 'ReservationStatusDate',
       'day_stay', 'foreigner', 'proportion', 'Country_simple', 'car_yes',
       'RA_changed'],
      dtype='object')

In [120]:
np.mean(df['RA_changed'])

0.18958985756136984

In [126]:
interested = ['IsCanceled', 'LeadTime', 'ArrivalDateYear', 'ArrivalDateMonth',
       'ArrivalDateWeekNumber', 'ArrivalDateDayOfMonth',
       'StaysInWeekendNights', 'StaysInWeekNights', 'Adults', 'Children',
       'Babies', 'Meal', 'Country', 'MarketSegment', 'DistributionChannel',
       'IsRepeatedGuest', 'PreviousCancellations',
       'PreviousBookingsNotCanceled', 'ReservedRoomType', 'AssignedRoomType',
       'BookingChanges', 'DepositType', 'Agent', 'Company',
       'DaysInWaitingList', 'CustomerType', 'ADR', 'RequiredCarParkingSpaces',
       'TotalOfSpecialRequests', 'ReservationStatus', 'ReservationStatusDate',
       'day_stay', 'foreigner', 'proportion', 'Country_simple', 'car_yes',
       'RA_changed']

df[interested].to_csv('test2_modifyed.csv', index=False)

In [ ]:
features = 